numpy/core/tests/test_array_coercion.py

```diff
@@ -0,0 +1,524 @@
+"""
+Tests for array coercion, mainly through testing `np.array` results directly.
+Note that other such tests exist e.g. in `test_api.py` and many corner-cases
+are tested (sometimes indirectly) elsewhere.
+"""
+
+import pytest
+
+from itertools import product
+
+import numpy as np
+from numpy.core._rational_tests import rational
+
+from numpy.testing import (
+    assert_array_equal, assert_warns, IS_PYPY)
+
+
+def arraylikes():
+    """
+    Generator for functions converting an array into various array-likes.
+    If full is True (default) includes array-likes not capable of handling
+    all dtypes
+    """
+    # base array:
+    def ndarray(a):
+        return a
+
+    yield ndarray
+
+    # subclass:
+    class MyArr(np.ndarray):
+        pass
+
+    def subclass(a):
+        return a.view(MyArr)
+
+    yield subclass
+
+    # Array-interface
+    class ArrayDunder:
+        def __init__(self, a):
+            self.a = a
+
+        def __array__(self, dtype=None):
+            return self.a
+
+    yield ArrayDunder
+
+    # memory-view
+    yield memoryview
+
+    # Array-interface
+    class ArrayInterface:
+        def __init__(self, a):
+            self.a = a  # need to hold on to keep interface valid
+            self.__array_interface__ = a.__array_interface__
+
+    yield ArrayInterface
+
+    # Array-Struct
+    class ArrayStruct:
+        def __init__(self, a):
+            self.a = a  # need to hold on to keep struct valid
+            self.__array_struct__ = a.__array_struct__
+
+    yield ArrayStruct
+
+
+def scalar_instances(times=True, extended_precision=True, user_dtype=True):
+    # Hard-coded list of scalar instances.
+    # Floats:
+    yield np.sqrt(np.float16(5))
+    yield np.sqrt(np.float32(5))
+    yield np.sqrt(np.float64(5))
+    if extended_precision:
+        yield np.sqrt(np.longdouble(5))
+
+    # Complex:
+    yield np.sqrt(np.complex64(2+3j))
+    yield np.sqrt(np.complex128(2+3j))
+    if extended_precision:
+        yield np.sqrt(np.longcomplex(2+3j))
+
+    # Integers:
+    yield np.int8(2)
+    yield np.int16(2)
+    yield np.int32(2)
+    yield np.int64(2)
+
+    yield np.uint8(2)
+    yield np.uint16(2)
+    yield np.uint32(2)
+    yield np.uint64(2)
```

Comment:
is there a reason np.bool8 not needed here or is it not necessary ?

Reply:
Thanks, seems I also forgot strings while refactoring...

Reply:
Argg, I commented it out for now, I had not thought of gh-9875 and my array-coercion PR of course homogenizes behaviour in that regard.  Although mainly for the numpy types, not for general strings.
We need to figure it out, either rip of the band-aid and fix it, or add on some type of band-aid before 1.20, for now I will open an issue for this and the bad float cases...